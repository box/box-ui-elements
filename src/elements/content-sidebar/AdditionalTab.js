/**
 * @flow
 * @file Preview sidebar additional tab component
 * @author Box
 */

import * as React from 'react';
import noop from 'lodash/noop';
import Tooltip from '../common/Tooltip';
import PlainButton from '../../components/plain-button/PlainButton';
import IconEllipsis from '../../icons/general/IconEllipsis';

import './AdditionalTab.scss';

const AdditionalTab = ({ callback: callbackFn = noop, iconUrl, id, title, ...rest }: AdditionalSidebarTab) => {
    return (
        <Tooltip position="middle-left" text={title}>
            <PlainButton
                className="bcs-additional-tab-btn bcs-nav-btn"
                type="button"
                onClick={() => callbackFn({ id, callbackData: rest })}
            >
                {id > 0 && iconUrl ? (
                    <img className="bcs-additional-tab-icon" src={iconUrl} alt={title} />
                ) : (
                    <IconEllipsis className="bcs-additional-tab-more-icon" />
                )}
            </PlainButton>
        </Tooltip>
    );
};

export default AdditionalTab;
