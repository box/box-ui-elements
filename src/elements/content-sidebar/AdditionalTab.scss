@import '../../styles/variables';

.bcs-additional-tab-icon {
    height: 24px;
    width: 24px;
}

.bcs-additional-tab-more-icon .fill-color {
    fill: $downtown-grey;
}
