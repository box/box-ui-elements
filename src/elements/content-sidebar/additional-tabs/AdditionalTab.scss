@import '../../../styles/variables';
@import '../../common/_animations';

.bcs-additional-tab-icon,
.bcs-additional-tab-more-icon {
    @include fade-in(.5s);

    height: 24px;
    width: 24px;
}

.bcs-additional-tab-more-icon .fill-color {
    fill: $downtown-grey;
}
