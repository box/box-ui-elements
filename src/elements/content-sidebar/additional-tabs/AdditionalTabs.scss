@import '../../../styles/variables';

.bdl-AdditionalTabs::before {
    background: $efive;
    content: '';
    display: block;
    height: 1px;
    margin: 16px auto;
    width: 44px;
}
