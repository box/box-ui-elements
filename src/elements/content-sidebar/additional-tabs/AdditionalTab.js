/**
 * @flow
 * @file Preview sidebar additional tab component
 * @author Box
 */

import * as React from 'react';
import noop from 'lodash/noop';
import classNames from 'classnames';
import Tooltip from '../../common/Tooltip';
import PlainButton from '../../../components/plain-button/PlainButton';
import IconEllipsis from '../../../icons/general/IconEllipsis';

import './AdditionalTab.scss';

const AdditionalTab = ({
    callback: callbackFn = noop,
    iconUrl,
    id,
    isLoading,
    onImageLoad,
    title,
    ...rest
}: AdditionalSidebarTab) => {
    const className = classNames('bcs-nav-btn', isLoading ? 'is-hidden' : '');

    return (
        <Tooltip position="middle-left" text={title}>
            <PlainButton className={className} type="button" onClick={() => callbackFn({ id, callbackData: rest })}>
                {id > 0 && iconUrl ? (
                    <img className="bcs-additional-tab-icon" src={iconUrl} onLoad={onImageLoad} alt={title} />
                ) : (
                    <IconEllipsis className="bcs-additional-tab-more-icon" />
                )}
            </PlainButton>
        </Tooltip>
    );
};

export default AdditionalTab;
