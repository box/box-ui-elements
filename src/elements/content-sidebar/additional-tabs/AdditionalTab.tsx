import * as React from 'react';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import Apps16 from '../../../icon/fill/Apps16';

import PlainButton from '../../../components/plain-button/PlainButton';
import AdditionalTabTooltip from './AdditionalTabTooltip';
import AdditionalTabPlaceholder from './AdditionalTabPlaceholder';
import messages from '../../common/messages';
import type { AdditionalSidebarTab, AdditionalSidebarTabFtuxData } from '../flowTypes';
import './AdditionalTab.scss';

interface Props extends AdditionalSidebarTab {
    ftuxTooltipData?: AdditionalSidebarTabFtuxData;
    isLoading: boolean;
    onImageLoad: () => void;
    status?: string;
}

const BLOCKED_BY_SHIELD = 'BLOCKED_BY_SHIELD_ACCESS_POLICY';

const AdditionalTab = ({
    callback: callbackFn,
    iconUrl,
    id,
    isLoading,
    onImageLoad,
    title,
    icon,
    ftuxTooltipData,
    status,
    ...rest
}: Props): React.ReactElement => {
    const [isErrored, setIsErrored] = React.useState<boolean>(false);

    const onImageError = (): void => {
        onImageLoad();
        setIsErrored(true);
    };

    const isDisabled = (): boolean => {
        return status === BLOCKED_BY_SHIELD;
    };

    const getDisabledReason = (): React.ReactNode => {
        let reason: React.ReactNode = null;
        switch (status) {
            case BLOCKED_BY_SHIELD:
                reason = <FormattedMessage {...messages.blockedByShieldAccessPolicy} />;
                break;
            default:
            // noop
        }
        return reason;
    };

    const getTabIcon = (): React.ReactNode => {
        let TabIcon;

        if (isErrored) {
            TabIcon = <AdditionalTabPlaceholder isLoading={false} />;
        } else if (id && id > 0) {
            TabIcon = (
                <img
                    className="bdl-AdditionalTab-icon"
                    src={iconUrl}
                    onError={onImageError}
                    onLoad={onImageLoad}
                    alt={title}
                />
            );
        } else {
            TabIcon = icon || <Apps16 className="bdl-AdditionalTab-icon" />;
        }

        return TabIcon;
    };

    const disabled = isDisabled();

    const className = classNames('bdl-AdditionalTab', {
        'bdl-is-hidden': isLoading,
        'bdl-is-disabled': disabled,
        'bdl-is-overflow': id && id < 0,
    });

    const tooltipText = disabled ? getDisabledReason() : title;

    return (
        <AdditionalTabTooltip
            defaultTooltipText={tooltipText}
            ftuxTooltipData={ftuxTooltipData}
            isFtuxVisible={!isLoading}
        >
            <PlainButton
                aria-label={title}
                className={className}
                data-testid="additionaltab"
                isDisabled={disabled}
                onClick={() => callbackFn({ id, callbackData: rest })}
            >
                {getTabIcon()}
            </PlainButton>
        </AdditionalTabTooltip>
    );
};

export default AdditionalTab;
