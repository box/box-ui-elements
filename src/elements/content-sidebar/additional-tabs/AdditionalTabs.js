/**
 * @flow
 * @file Preview sidebar additional tabs component
 * @author Box
 */

import React, { PureComponent } from 'react';
import AdditionalTab from './AdditionalTab';
import AdditionalTabsLoading from './AdditionalTabsLoading';

import './AdditionalTabs.scss';

type Props = {
    tabs?: Array<AdditionalSidebarTab>,
};

type State = {
    isLoading: boolean,
};

class AdditionalTabs extends PureComponent<Props, State> {
    numLoadedTabs: number = 0;

    constructor(props: Props) {
        super(props);

        this.state = { isLoading: true };
    }

    componentDidMount() {
        this.calculateNumberOfLoadableTabs();
    }

    componentDidUpdate() {
        this.calculateNumberOfLoadableTabs();
    }

    /**
     * Determines if we need to account for a "more" tab when loading icon images.
     *
     * @return {void}
     */
    calculateNumberOfLoadableTabs() {
        const { tabs } = this.props;

        if (tabs) {
            // If we're displaying the more options tab, consider it loaded since it doesn't use an image.
            const hasMoreTab = tabs.find(tab => tab.id < 0 && !tab.iconUrl);
            const moreTabCount = hasMoreTab ? 1 : 0;

            this.numLoadedTabs = moreTabCount;
        }
    }

    /**
     * Handles individual icon image load
     *
     * @return {void}
     */
    onImageLoad = () => {
        const { tabs } = this.props;

        const numTabs = tabs ? tabs.length : 0;
        this.numLoadedTabs += 1;

        if (this.numLoadedTabs === numTabs) {
            this.setState({
                isLoading: false,
            });
        }
    };

    render() {
        const { tabs } = this.props;
        const { isLoading } = this.state;

        return (
            <div className="bdl-AdditionalTabs">
                {isLoading && <AdditionalTabsLoading />}
                {tabs &&
                    tabs.map(tabData => (
                        <AdditionalTab
                            key={tabData.id}
                            onImageLoad={this.onImageLoad}
                            isLoading={isLoading}
                            {...tabData}
                        />
                    ))}
            </div>
        );
    }
}

export default AdditionalTabs;
