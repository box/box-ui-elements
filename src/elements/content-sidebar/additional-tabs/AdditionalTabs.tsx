import React, { PureComponent } from 'react';
import AdditionalTab from './AdditionalTab';
import AdditionalTabsLoading from './AdditionalTabsLoading';
import { AdditionalTabsProps } from './types';
import './AdditionalTabs.scss';

type State = {
    isLoading: boolean;
};

class AdditionalTabs extends PureComponent<AdditionalTabsProps, State> {
    numLoadedTabs = 0;

    constructor(props: AdditionalTabsProps) {
        super(props);
        this.state = { isLoading: true };
    }

    /**
     * Handles an individual icon image load
     */
    onImageLoad = () => {
        const { tabs } = this.props;

        if (!tabs) {
            return;
        }

        const hasMoreTab = tabs.find(tab => tab.id < 0 && !tab.iconUrl);
        const numTabs = tabs.length - (hasMoreTab ? 1 : 0);

        this.numLoadedTabs += 1;

        if (this.numLoadedTabs === numTabs) {
            this.setState({
                isLoading: false,
            });
        }
    };

    render() {
        const { tabs } = this.props;
        const { isLoading } = this.state;

        return (
            <div className="bdl-AdditionalTabs">
                {isLoading && <AdditionalTabsLoading />}
                {tabs &&
                    tabs.map(tabData => (
                        <AdditionalTab
                            key={tabData.id}
                            onImageLoad={this.onImageLoad}
                            isLoading={isLoading}
                            {...tabData}
                        />
                    ))}
            </div>
        );
    }
}

export default AdditionalTabs;
