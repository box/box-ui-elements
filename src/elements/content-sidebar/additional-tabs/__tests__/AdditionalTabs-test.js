import React from 'react';
import { shallow } from 'enzyme';
import AdditionalTabs from '../AdditionalTabs';
import AdditionalTab from '../AdditionalTab';

describe('elements/content-sidebar/AdditionalTabs', () => {
    const getWrapper = props => shallow(<AdditionalTabs {...props} />);

    test('should render the correct number of tabs', () => {
        const props = {
            tabs: [
                {
                    id: 200,
                    title: 'Test title',
                    iconUrl: 'https://foo.com/icon',
                    callback: jest.fn(),
                    status: 'ADDED',
                },
                {
                    id: 1,
                    title: 'Another title',
                    iconUrl: 'https://foo.com/icon',
                    callback: jest.fn(),
                    status: 'ADDED',
                },
            ],
        };

        const wrapper = getWrapper(props);
        expect(wrapper.find(AdditionalTab)).toHaveLength(2);
        expect(wrapper).toMatchSnapshot();
    });
});
