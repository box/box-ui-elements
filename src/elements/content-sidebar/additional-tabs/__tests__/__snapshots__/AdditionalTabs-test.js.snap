// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/content-sidebar/AdditionalTabs should render the correct number of tabs 1`] = `
<div
  className="bcs-nav-additional-tabs"
>
  <AdditionalTabsLoading />
  <AdditionalTab
    callback={[MockFunction]}
    iconUrl="https://foo.com/icon"
    id={200}
    isLoading={true}
    key="200"
    onImageLoad={[Function]}
    status="ADDED"
    title="Test title"
  />
  <AdditionalTab
    callback={[MockFunction]}
    iconUrl="https://foo.com/icon"
    id={1}
    isLoading={true}
    key="1"
    onImageLoad={[Function]}
    status="ADDED"
    title="Another title"
  />
</div>
`;
